---
title: "Comparing methods on Schraivogel discovery data"
author: "Gene"
date: "2023-04-18"
output: pdf_document
---

```{r, message = FALSE, echo = FALSE}
library(readr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(cowplot)
library(kableExtra)
library(conflicted)
conflicted::conflicts_prefer(dplyr::filter)

sceptre2_dir <- .get_config_path("LOCAL_SCEPTRE2_DATA_DIR")
schraivogel_data_dir <- .get_config_path("LOCAL_SCHRAIVOGEL_2020_DATA_DIR")
result_sceptre <- readRDS(
  paste0(sceptre2_dir, 
         "results/schraivogel_analysis/sceptre_schraivogel_chr_8_results.rds")
)
result_schraivogel <- read_csv(
  paste0(schraivogel_data_dir, "raw/ftp/chromatin_annotated_pairs.csv")
)
result_seurat <- readRDS(paste0(
  sceptre2_dir, 
  "results/discovery_analyses/seurat_schraivogel_enhancer_screen_chr8_gene_res.rds"))

replace_periods <- function(str){
  stringr::str_replace(stringr::str_replace(str, "[.]", ":"), "[.]", "-")
}

q <- 0.1
result_merged <- result_schraivogel |>
  mutate(perturbation = replace_periods(perturbation)) |>
  filter(sample == "chr8", grepl("chr", perturbation)) |>
  rename(grna_group = perturbation, response_id = gene, pvalue_schraivogel = pvalue) |>
  select(response_id, grna_group, pvalue_schraivogel, dist_to_tss, int_freq) |>
  left_join(result_sceptre |> 
              rename(pvalue_sceptre = p_value) |> 
              select(response_id, grna_group, pvalue_sceptre), 
            by = c("response_id", "grna_group")) |>
  left_join(result_seurat |>
              rename(pvalue_seurat = p_value), 
            by = c("response_id", "grna_group")) |>
  mutate(pvalue_all = 0) |>
  na.omit() |> 
  pivot_longer(starts_with("pvalue_"), 
               names_to = "method", 
               values_to = "pvalue", 
               names_prefix = "pvalue_") |>
  group_by(method) |> 
  mutate(discovery = p.adjust(pvalue, "fdr") <= q) |>
  ungroup() |>
  mutate(method = factor(method, 
                         levels = c("schraivogel", "sceptre", "seurat", "all"),
                         labels = c("Schraivogel", "SCEPTRE", "Seurat", "All pairs")))
```

In this writeup, I will compare the SCEPTRE (exact), Seurat DE, and the Schraivogel method (using 1000 randomly-selected cells from the complement set as the control group) on the Schraivogel discovery data from chromosome 8. I subject the p-values generated by each method to the BH correction at target level 0.1.

# Numbers of discoveries

```{r, echo = FALSE}
result_merged |> 
  filter(discovery, method != "all") |>
  group_by(method) |>
  summarise(`Number of discoveries` = n()
#            `Mean dist` = mean(dist_to_tss),
#            `Median dist` = median(dist_to_tss),
#            `Mean int freq` = mean(int_freq),
#            `Median int freq` = median(int_freq)
) |>
  rename(Method = method) |>
  kable(format = "latex", booktabs = TRUE) |>
  kable_styling(position = "center", latex_options = "HOLD_position")
```

SCEPTRE and Schraivogel make a few dozen discoveries each, whereas Seurat discovers about a quarter of all pairs tested.

# Enhancer-gene distance

```{r, fig.width = 4, fig.height = 3, fig.align='center', echo = FALSE}
result_merged |> 
  filter(discovery) |>
  ggplot(aes(x = method, y = dist_to_tss/1e3)) +
  geom_violin(scale = "width", width = 0.5, fill = "dodgerblue") +
  geom_boxplot(width = 0.1, outlier.shape = NA, fill = "dodgerblue") +
  scale_y_log10() +
  labs(x = NULL, 
       y = "Enhancer-gene distance (kb)") 
```

Schraivogel has the smallest gene-enhancer distances, followed by SCEPTRE, followed distantly by Seurat.

# HI-C interaction frequency

```{r, fig.width = 4, fig.height = 3, echo = FALSE, fig.align='center'}
result_merged |> 
  filter(discovery) |>
  ggplot(aes(x = method, y = int_freq)) +
  geom_violin(scale = "width", width = 0.5, fill = "dodgerblue") +
  geom_boxplot(width = 0.1, outlier.shape = NA, fill = "dodgerblue")  +
  scale_y_continuous(trans = "log1p", breaks = c(0, 1, 3, 10, 30, 100)) +
  labs(x = NULL, 
       y = "HI-C interaction frequency") 
```

Schraivogel has the largest HI-C interaction frequencies, followed by SCEPTRE, followed distantly by Seurat.

# Comparing SCEPTRE and Schraivogel p-values

```{r, echo = FALSE, fig.width = 3.5, fig.height = 3, fig.align='center'}
schraivogel_thresh <- result_merged |>
  filter(method == "Schraivogel", discovery) |>
  summarise(max(pvalue)) |>
  pull()

sceptre_thresh <- result_merged |>
  filter(method == "SCEPTRE", discovery) |>
  summarise(max(pvalue)) |>
  pull()

result_merged |>
  select(-discovery) |>
  mutate(pvalue = pmax(1e-16, pvalue)) |>
  pivot_wider(names_from = method, values_from = pvalue) |>
  filter(SCEPTRE < 0.05 | Schraivogel < 0.05 | row_number() %% 25 == 0) |>
  ggplot(aes(x = SCEPTRE, y = Schraivogel)) +
  geom_point() +
  geom_vline(xintercept = sceptre_thresh, linetype = "dashed") +
  geom_hline(yintercept = schraivogel_thresh, linetype = "dashed") +
  geom_abline() +
  scale_x_log10() +
  scale_y_log10() + 
  labs(x = "SCEPTRE p-value",
       y = "Schraivogel p-value")
```

The SCEPTRE and Schraivogel p-values are not particularly well aligned, although there is nontrivial overlap between the corresponding discovery sets.

# Conclusion

The Schraivogel method is clearly the best. The SCEPTRE discoveries are clearly enriched for biological signal, but not as much so as Schraivogel's. The reason for Schraivogel's superior performance might be that it sidesteps the weird confounding issue by using the complement set rather than the NTC set as controls. Seurat is the worst, its discoveries not being enriched for biological signal. The reason for this is that it does not control for batch but uses the NTC set as controls, so it suffers from the extreme confounding issue.