---
title: "TF SCEPTRE vs Seurat With Score QC Summary"
output: pdf_document
date: "2023-04-04"
header-includes:
- \usepackage{diagbox}
- \usepackage{multicol}
- \usepackage{float}
---

```{r, echo = FALSE,message = F}
knitr::opts_chunk$set(fig.align = 'center', echo = FALSE)
```

```{r, message = FALSE, warning = FALSE, cache = FALSE,echo = FALSE}
# load libraries
library(tidyverse)
library(plyranges)
library(genomation)
library(biomaRt)
library(GenomicRanges)
library(kableExtra)

# resolve namespace conflicts
library(conflicted)
conflicts_prefer(dplyr::filter)
conflicts_prefer(dplyr::select)
conflicts_prefer(dplyr::rename)
conflicts_prefer(base::intersect)
```


```{r,echo = FALSE}
score_QC = 0.9
data.dir = .get_config_path("LOCAL_SCEPTRE2_DATA_DIR")
```



```{r, message = FALSE,warning = FALSE,echo = FALSE}
data_dir <- .get_config_path("LOCAL_SCEPTRE2_DATA_DIR")

# read in sceptre and seurat results
sceptre_path <- paste0(
  data_dir, "results/discovery_analyses/",
  "papalexi_gene_discovery_res.rds"
)
seurat_path <- paste0(
  data_dir, "results/papalexi_analysis/",
  "seurat_all_perturbations_results.rds"
)
sceptre_results <- readRDS(sceptre_path)
seurat_results <- readRDS(seurat_path)

atac_path = paste0(data_dir, "results/papalexi_analysis/",
  "ATACseq_TF_targets_score_QC_",100*score_QC,".rds")

atac_target_genes = readRDS(atac_path)
```







```{r,message =F,warning = FALSE,echo = FALSE}
TFs = colnames(atac_target_genes)
odds_ratios <- matrix(NA, nrow = length(TFs), ncol = 7, 
                      dimnames = list(ground_truth = TFs, 
          method = c("SCEPTRE", "Seurat","SCEPTRE-pval","Seurat-pval",
                "Num TF Targets","Num SCEPTRE Targets","Num Seurat Targets")))

for(TF in TFs){
  #get atac target gene matrix
  atac_TF = atac_target_genes%>%select(TF)%>%
    mutate(gene = rownames(atac_target_genes))%>%
    dplyr::rename(binary_score = TF)
  
  atac_genes = rownames(atac_target_genes)
  genes_qc <- intersect(sceptre_results |> 
                          filter(grna_group == TF) |>
                          na.omit() |>
                          pull(response_id),
                        atac_genes)
  
  q <- 0.1
  
  # apply BH correction to SCEPTRE STAT1 results
  sceptre_results_TF<- sceptre_results |>
    filter(grna_group == TF, response_id %in% genes_qc) |>
    mutate(sceptre_discovery = p.adjust(p_value, method = "BH") <= q) |>
    select(response_id, sceptre_discovery) |>
    rename(gene = response_id)
  
  # apply BH correction to Seurat TFresults
  seurat_results_TF<- seurat_results |>
    filter(grna_group == TF, response_id %in% genes_qc) |>
    mutate(seurat_discovery = p.adjust(p_val, method = "BH") <= q) |>
    select(response_id, seurat_discovery) |>
    rename(gene = response_id)
  
  # join with chipseq scores
  merged_results <- atac_TF |>
    filter(gene %in% genes_qc) |>
    left_join(sceptre_results_TF, by = "gene") |>
    left_join(seurat_results_TF, by = "gene")
  
  
  
  odds_ratios[TF, "SCEPTRE"] <- merged_results |> 
    select(binary_score, sceptre_discovery) |>
    table() |>
    fisher.test() %>%
    `$`("estimate")
  
  odds_ratios[TF, "Seurat"] <- merged_results |> 
    select(binary_score, seurat_discovery) |>
    table() |>
    fisher.test() %>%
    `$`("estimate")
  
  odds_ratios[TF, "SCEPTRE-pval"] <- merged_results |> 
    select(binary_score, sceptre_discovery) |>
    table() |>
    fisher.test() %>%
    `$`("p.value")
  
  odds_ratios[TF, "Seurat-pval"] <- merged_results |> 
    select(binary_score, seurat_discovery) |>
    table() |>
    fisher.test() %>%
    `$`("p.value")
  
  odds_ratios[TF, "Num TF Targets"] <- sum(merged_results[["binary_score"]])
  
  
  odds_ratios[TF, "Num Seurat Targets"] <- sum(merged_results[[
    "seurat_discovery"]])
  
  
  odds_ratios[TF, "Num SCEPTRE Targets"] <- sum(merged_results[[
    "sceptre_discovery"]])
  
  
}


```



```{r,results = 'asis',message = F,echo = FALSE}
odds_ratios |>
  signif(3) |>
  kable(booktabs = TRUE,caption = "Enrichment odds ratios, comparing to ATAC-seq target assignments with changing quantile threshold")%>%
  kable_styling(latex_options = "scale_down")

```


\newpage
