---
title: "TF SCEPTRE vs Seurat in monocytes, with same QC"
output: pdf_document
date: "2023-04-04"
header-includes:
- \usepackage{diagbox}
- \usepackage{multicol}
- \usepackage{float}
---

```{r, echo = FALSE}
knitr::opts_chunk$set(fig.align = 'center', echo = FALSE)
```

```{r, message = FALSE, warning = FALSE, cache = FALSE}
# load libraries
library(tidyverse)
library(plyranges)
library(genomation)
library(biomaRt)
library(GenomicRanges)
library(kableExtra)

# resolve namespace conflicts
library(conflicted)
conflicts_prefer(dplyr::filter)
conflicts_prefer(dplyr::select)
conflicts_prefer(dplyr::rename)
conflicts_prefer(base::intersect)
```

# Introduction

Here we analyze the performance of SCEPTRE and Seurat on the broad condition gene sets from the HTFtarget database. You can input the TF you wan tto analyze in the code chunk below.

```{r,echo = T}
#get the TF name 
#TF_names = c("STAT1","STAT2","BRD4","MYC","SMAD4","IRF1")
TF_name = "SMAD4"

```

```{r, message = FALSE}
data_dir <- .get_config_path("LOCAL_SCEPTRE2_DATA_DIR")

# read in sceptre and seurat results
sceptre_path <- paste0(
  data_dir, "results/papalexi_analysis/",
  "sceptre_full_mrna_results_with_effect_size.rds"
)
seurat_path <- paste0(
  data_dir, "results/papalexi_analysis/",
  "seurat_all_perturbations_results.rds"
)
sceptre_results <- readRDS(sceptre_path)
seurat_results <- readRDS(seurat_path)

# read in K562 ChIP-seq data
TF_filename <- "GSM1057025_IRF1peak_B.txt"
chipseq_fp <- paste0(data_dir, "data/chipseq/", TF_filename)
chipseq_data <- read.table(chipseq_fp) 
colnames(chipseq_data) = c("chrom",'start_pos','end_pos',"pval","score",
                           "pos_max_peak","max_peak_height",
                           "rel_pos_max_peak_height","peak_size","mid_point",
                           "peak_to_mid_dist")

# read in genes from database
database_genes <- read_table(paste0(
  data_dir,
  "data/htftarget/",TF_name,".target.txt"
)) |>
  suppressWarnings() |>
  pull(target_name)

```

```{r}
chipseq_data = GRanges(
  seqnames = chipseq_data$chrom,
  ranges = IRanges(start = chipseq_data$start_pos, end = chipseq_data$end_pos),
  score = chipseq_data$score,
  peak_position = chipseq_data$pos_max_peak)
```

```{r}
# get TSS for each gene 
gene_names <- sceptre_results |> pull(response_id) |> unique() |> as.character()
ensembl <- useEnsembl(host = 'https://grch37.ensembl.org',
                      biomart = 'ENSEMBL_MART_ENSEMBL', 
                      dataset = "hsapiens_gene_ensembl")
TSS_info <-getBM(attributes=c("external_gene_name", "chromosome_name", "start_position", 
                       "end_position", "strand"),
                 filters=c('external_gene_name'),
                 value = gene_names, mart=ensembl) |>
  filter(chromosome_name %in% c(1:22, "X", "Y")) |>
  mutate(TSS = ifelse(strand == 1, start_position, end_position),
         chromosome_name = paste0("chr", chromosome_name)) |>
  rename(gene = external_gene_name, chr = chromosome_name) |>
  select(gene, chr, TSS)
```


```{r}
# get binary score for each gene
window_width_binary <- 5e3
chipseq_scores_binary <- GRanges(
  seqnames = TSS_info$chr,
  ranges = IRanges(start = TSS_info$TSS-window_width_binary, 
                   end = TSS_info$TSS+window_width_binary),
  gene = TSS_info$gene,
  TSS = TSS_info$TSS) |> 
  join_overlap_left(chipseq_data) |>
  group_by(gene) |>
  summarise(binary_score = any(!is.na(score))) |>
  as_tibble()

# get linear score for each gene
window_width_linear <- 5e4
chipseq_scores_linear <- GRanges(
  seqnames = TSS_info$chr,
  ranges = IRanges(start = TSS_info$TSS-window_width_linear, 
                   end = TSS_info$TSS+window_width_linear),
  gene = TSS_info$gene,
  TSS = TSS_info$TSS) |> 
  join_overlap_left(chipseq_data) |>
  group_by(gene) |>
  summarise(linear_score = sum((window_width_linear - abs(TSS - peak_position))/
                                 window_width_linear)) |>
  as_tibble() |>
  mutate(linear_score = ifelse(is.na(linear_score), 0, linear_score))
```

```{r}
# join the two sets of chipseq scores, and add database information
chipseq_scores <- full_join(chipseq_scores_binary, 
                            chipseq_scores_linear, 
                            by = "gene") |>
  mutate(in_database = gene %in% database_genes)
```


```{r, fig.width = 4, fig.height = 3, eval = FALSE}
chipseq_scores |> 
  group_by(in_database) |>
  summarise(round(mean(binary_score), 3)) |>
  kable(booktabs = TRUE, 
        linesep = "",
        col.names = c("In database", "Mean binary score"),
        caption = paste0("Comparing ChIP-seq binary score to database.")) |>
  kable_styling(position = "center", latex_options = "HOLD_position")

chipseq_scores |> 
  group_by(in_database) |>
  summarise(round(median(linear_score), 3)) |>
  kable(booktabs = TRUE, 
        linesep = "",
        col.names = c("In database", "Median linear score"),
        caption = paste0("Comparing ChIP-seq linear score to database.")) |>
  kable_styling(position = "center", latex_options = "HOLD_position")

chipseq_scores |> 
  ggplot(aes(x = in_database, y = linear_score)) +
  geom_violin(fill = "dodgerblue") +
  labs(x = "In database", y = "Linear score")
```

# Compare SCEPTRE and Seurat results with ChIP-seq scores

Let's first look at how the SCEPTRE and Seurat discoveries align with each other.
```{r}
genes_qc <- intersect(sceptre_results |> 
                        filter(grna_group == TF_name) |>
                        na.omit() |>
                        pull(response_id),
                      chipseq_scores |> pull(gene))

q <- 0.1

# apply BH correction to SCEPTRE TF results
sceptre_results_TF <- sceptre_results |>
  filter(grna_group == TF_name, response_id %in% genes_qc) |>
  mutate(sceptre_discovery = p.adjust(p_value, method = "BH") <= q) |>
  select(response_id, sceptre_discovery) |>
  rename(gene = response_id)

# apply BH correction to Seurat TF results
seurat_results_TF <- seurat_results |>
  filter(grna_group == TF_name, response_id %in% genes_qc) |>
  mutate(seurat_discovery = p.adjust(p_val, method = "BH") <= q) |>
  select(response_id, seurat_discovery) |>
  rename(gene = response_id)

# join with chipseq scores
merged_results <- chipseq_scores |>
  filter(gene %in% genes_qc) |>
  left_join(sceptre_results_TF, by = "gene") |>
  left_join(seurat_results_TF, by = "gene")
```

```{r, results='asis'}
merged_results |> 
  select(sceptre_discovery, seurat_discovery) |>
  table() |>
  Hmisc::latex(booktabs = TRUE,
               rowlabel="\\diagbox{SCEPTRE}{Seurat}", 
               file="",
               caption = "Comparing SCEPTRE versus Seurat discoveries.",
               where = "!htbp")
```

This table suggests that SCEPTRE and Seurat results have decent, but imperfect, agreement. Also note that the total numbers of discoveries made by the two methods are nearly the same. Next, let's look at how the SCEPTRE and Seurat discoveries align with the ChIP-seq target genes (as identified by either the hTFtarget database or directly from the ChIP-seq data).

\newpage

```{r, results='asis'}
add_proportions <- function(tab){
  tab |>
    addmargins(quiet = TRUE, FUN = list(Prop = function(x){
      if(any(x < 1)) NA_character_ else round(x["TRUE"]/sum(x), 3)
    }))
}

merged_results |> 
  select(in_database, sceptre_discovery) |>
  table() |>
  add_proportions() |>
  Hmisc::latex(booktabs = TRUE,
             rowlabel="\\diagbox{Database}{SCEPTRE}", 
             file="",
             caption = "Comparing SCEPTRE to database.", 
             where = "!htbp")

merged_results |> 
  select(in_database, seurat_discovery) |>
  table() |>
  add_proportions() |>
  Hmisc::latex(booktabs = TRUE,
           rowlabel="\\diagbox{Database}{Seurat}", 
           file="",
           caption = "Comparing Seurat to database.",
           where = "!htbp")

```

The proportions are the proportion of `TRUE` values in each row or column. For example, 0.576 of the genes found by SCEPTRE are marked as TF targets in the database (i.e. SCEPTRE has specificity 0.576). From these tables, we see that **SCEPTRE has slightly lower sensitivity and specificity than Seurat.** We can summarize each 2-by-2 table via its odds ratio (the p-values are all extremely small). The resulting odds ratios are shown below.

```{r, results = 'asis'}
odds_ratios <- matrix(NA, nrow = 2, ncol = 2, 
                      dimnames = list(ground_truth = c("database","pvalue"), 
                                      method = c("SCEPTRE", "Seurat")))

odds_ratios["database", "SCEPTRE"] <- merged_results |> 
  select(in_database, sceptre_discovery) |>
  table() |>
  fisher.test() %>%
  `$`("estimate")

odds_ratios["database", "Seurat"] <- merged_results |> 
  select(in_database, seurat_discovery) |>
  table() |>
  fisher.test() %>%
  `$`("estimate")

odds_ratios["pvalue", "SCEPTRE"] <- merged_results |> 
  select(in_database, sceptre_discovery) |>
  table() |>
  fisher.test() %>%
  `$`("p.value")

odds_ratios["pvalue", "Seurat"] <- merged_results |> 
  select(in_database, seurat_discovery) |>
  table() |>
  fisher.test() %>%
  `$`("p.value")




odds_ratios |>
  round(3) |>
  Hmisc::latex(booktabs = TRUE,
           rowlabel="\\diagbox{Ground truth}{Method}", 
           file="",
           caption = "Enrichment odds ratios, comparing to database and our ChIP-seq target assignments.",
           where = "!htbp")
```

