---
title: "SCEPTRE vs Seurat ChIP-seq Analysis for IRF1"
author: "Gene"
output: pdf_document
date: "2023-04-02"
urlcolor: blue
---

```{r, echo = FALSE}
knitr::opts_chunk$set(fig.align = 'center', echo = FALSE, cache = TRUE)
```

```{r, message = FALSE, warning = FALSE, cache = FALSE}
# load libraries
library(tidyverse)
library(plyranges)
library(genomation)
library(biomaRt)
library(GenomicRanges)
library(kableExtra)

# resolve namespace conflicts
library(conflicted)
conflicts_prefer(dplyr::filter)
conflicts_prefer(dplyr::select)
conflicts_prefer(dplyr::rename)
```

# Introduction

The goal of this writeup is to take a closer look at the ChIP-seq enrichment analysis, focusing on IRF1. In particular, I examine the following questions:

- How do people typically determine transcription factor targets?
- How do ChIP-seq scores for potential target genes align with those in the hTFtarget database?
- How do SCEPTRE and Seurat IRF1 discoveries align with each other? 
- How do SCEPTRE and Seurat IRF1 discoveries align with ChIP-seq scores?

# Determining transcription factor targets

[Sikora-Wohlfeld et al, 2013](https://journals.plos.org/ploscompbiol/article/file?id=10.1371/journal.pcbi.1003342&type=printable) discuss various computational methods to infer TF targets based on ChIP-seq data. This problem turns out to be not completely straightforward, with several competing methodologies available. The one we had been trying, which Sikora-Wohlfeld et al call the "binary" approach, is the most naive method. It has fairly poor performance. However, Sikora-Wohlfeld find that the best window width to use for this approach is 5kb, and they consider an interval of this width centered on the TSS rather than just upstream of it. An approach that performs better is called the "linear" approach. In this approach, the relative distances of ChIP-seq peaks to the TSS are summed, restricting attention to a 50kb window centered on the TSS. Another approach is [TIP](https://academic.oup.com/bioinformatics/article/27/23/3221/234205), which is the preferred approach of [ENCODE](https://www.nature.com/articles/nature11245). Unfortunately, the [TIP results for ENCODE](http://encodenets.gersteinlab.org/) are available only for K562 and GM12878 cell types (probably without IFN-gamma stimulation), and the [TIP software](http://archive.gersteinlab.org/proj/tftarget/) appears somewhat out of date. Hence, in this writeup, I will focus on the binary and linear scoring approaches outlined by Sikora-Wohlfeld et al.

```{r, message = FALSE}
data_dir <- .get_config_path("LOCAL_SCEPTRE2_DATA_DIR")

# read in sceptre and seurat results
sceptre_path <- paste0(
  data_dir, "results/papalexi_analysis/",
  "sceptre_full_mrna_results_with_effect_size.rds"
)
seurat_path <- paste0(
  data_dir, "results/papalexi_analysis/",
  "seurat_all_perturbations_results.rds"
)
sceptre_results <- readRDS(sceptre_path)
seurat_results <- readRDS(seurat_path)

# read in K562 ChIP-seq data
irf1_filename <- "GSM935549_hg19_wgEncodeSydhTfbsK562Irf1Ifng6hStdPk.narrowPeak"
chipseq_fp <- paste0(data_dir, "data/chipseq/", irf1_filename)
chipseq_data <- readNarrowPeak(chipseq_fp, track.line = FALSE, zero.based = TRUE) |>
  mutate(peak_position = round(0.5 * (start + end)))

# read in genes from database
database_genes <- read_table(paste0(
  data_dir,
  "data/htftarget/dataset_1762.IRF1.target.txt"
)) |>
  suppressWarnings() |>
  pull(target_name)
```

# Get ChIP-seq scores for each gene

```{r}
# get TSS for each gene 
gene_names <- sceptre_results |> pull(response_id) |> unique() |> as.character()
ensembl <- useEnsembl(host = 'https://grch37.ensembl.org',
                      biomart = 'ENSEMBL_MART_ENSEMBL', 
                      dataset = "hsapiens_gene_ensembl")
TSS_info <-getBM(attributes=c("hgnc_symbol", "chromosome_name", "start_position", 
                       "end_position", "strand"),
                 filters=c('hgnc_symbol'),
                 value = gene_names, mart=ensembl) |>
  filter(chromosome_name %in% c(1:22, "X", "Y")) |>
  mutate(TSS = ifelse(strand == 1, start_position, end_position),
         chromosome_name = paste0("chr", chromosome_name)) |>
  rename(gene = hgnc_symbol, chr = chromosome_name) |>
  select(gene, chr, TSS)
```


```{r, echo = TRUE}
# get binary score for each gene
window_width_binary <- 5e3
chipseq_scores_binary <- GRanges(
  seqnames = TSS_info$chr,
  ranges = IRanges(start = TSS_info$TSS-window_width_binary, 
                   end = TSS_info$TSS+window_width_binary),
  gene = TSS_info$gene,
  TSS = TSS_info$TSS) |> 
  join_overlap_left(chipseq_data) |>
  group_by(gene) |>
  summarise(binary_score = any(!is.na(score))) |>
  as_tibble()

# get linear score for each gene
window_width_linear <- 5e4
chipseq_scores_linear <- GRanges(
  seqnames = TSS_info$chr,
  ranges = IRanges(start = TSS_info$TSS-window_width_linear, 
                   end = TSS_info$TSS+window_width_linear),
  gene = TSS_info$gene,
  TSS = TSS_info$TSS) |> 
  join_overlap_left(chipseq_data) |>
  group_by(gene) |>
  summarise(linear_score = sum((window_width_linear - abs(TSS - peak_position))/
                                 window_width_linear)) |>
  as_tibble() |>
  mutate(linear_score = ifelse(is.na(linear_score), 0, linear_score))
```

```{r}
# join the two sets of chipseq scores, and add database information
chipseq_scores <- full_join(chipseq_scores_binary, 
                            chipseq_scores_linear, 
                            by = "gene") |>
  mutate(in_database = gene %in% database_genes)
```

Let's take a look at the distribution of the linear ChIP-seq scores. 
```{r, fig.width = 4, fig.height = 3, message = FALSE}
chipseq_scores |> 
  ggplot(aes(x = linear_score)) +
  geom_histogram(color = "black") +
  labs(x = "ChIP-seq linear score") +
  theme(axis.title.y = element_blank())
```

We see that there are modes at 0 (no peaks within the window width) and 1 (one peaks near the TSS). There is also a long right tail. Next, we check how these ChIP-seq scores align with the hTFtarget database.

```{r, fig.width = 4, fig.height = 3}
chipseq_scores |> 
  group_by(in_database) |>
  summarise(round(mean(binary_score), 2)) |>
  kable(booktabs = TRUE, 
        linesep = "",
        col.names = c("In database", "Mean binary score"),
        caption = paste0("Comparing ChIP-seq binary score to database.")) |>
  kable_styling(position = "center", latex_options = "hold_position")

chipseq_scores |> 
  group_by(in_database) |>
  summarise(round(median(linear_score), 2)) |>
  kable(booktabs = TRUE, 
        linesep = "",
        col.names = c("In database", "Median linear score"),
        caption = paste0("Comparing ChIP-seq linear score to database.")) |>
  kable_styling(position = "center", latex_options = "hold_position")

chipseq_scores |> 
  ggplot(aes(x = in_database, y = linear_score)) +
  geom_violin(fill = "dodgerblue") +
  labs(x = "In database", y = "Linear score")
```

These results suggest that there is decent agreement between the hTFtarget database and the scores we derived from the ChIP-seq data.

# Compare SCEPTRE and Seurat results with ChIP-seq scores

Let's first look at how the SCEPTRE and Seurat discoveries align with each other.
```{r}
q <- 0.1

# apply BH correction to SCEPTRE IRF1 results
sceptre_results_IRF1 <- sceptre_results |>
  filter(grna_group == "IRF1") |>
  mutate(sceptre_discovery = p.adjust(p_value, method = "BH") <= q) |>
  select(response_id, sceptre_discovery) |>
  rename(gene = response_id)

# apply BH correction to Seurat IRF1 results
seurat_results_IRF1 <- seurat_results[["IRF1"]] |>
  rownames_to_column(var = "gene") |>
  mutate(seurat_discovery = p.adjust(p_val, method = "BH") <= q) |>
  select(gene, seurat_discovery)

# join with chipseq scores
merged_results <- chipseq_scores |>
  left_join(sceptre_results_IRF1, by = "gene") |>
  left_join(seurat_results_IRF1, by = "gene") |>
  na.omit()
```

```{r}
merged_results |> 
  count(sceptre_discovery, seurat_discovery) |>
  kable(booktabs = TRUE, 
        linesep = "",
        col.names = c("SCEPTRE discovery", "Seurat Discovery", "Number"),
        caption = paste0("Comparing numbers of discoveries made by SCEPTRE 
                         and Seurat.")) |>
  kable_styling(position = "center", latex_options = "hold_position")
```

This table suggests that SCEPTRE and Seurat results have decent, but highly imperfect, agreement. Next, let's look at how the SCEPTRE and Seurat discoveries align with the ChIP-seq signal (as measured by the hTFtarget database, the binary ChIP-seq scores, and the linear ChIP-seq scores).

```{r}
merged_results |> 
  group_by(in_database) |>
  summarise(round(mean(sceptre_discovery), 2)) |>
  kable(booktabs = TRUE, 
      linesep = "",
      col.names = c("In hTFtarget", "Proportion SCEPTRE discoveries"),
      caption = paste0("Comparing SCEPTRE discoveries to ChIP-seq binary scores.")) |>
  kable_styling(position = "center", latex_options = "hold_position")

merged_results |> 
  group_by(in_database) |>
  summarise(round(mean(seurat_discovery), 2)) |>
  kable(booktabs = TRUE, 
      linesep = "",
      col.names = c("In hTFtarget", "Proportion Seurat discoveries"),
      caption = paste0("Comparing Seurat discoveries to ChIP-seq binary scores.")) |>
  kable_styling(position = "center", latex_options = "hold_position")
```

```{r}
merged_results |> 
  group_by(binary_score) |>
  summarise(round(mean(sceptre_discovery), 2)) |>
  kable(booktabs = TRUE, 
      linesep = "",
      col.names = c("Binary ChIP-seq score", "Proportion SCEPTRE discoveries"),
      caption = paste0("Comparing SCEPTRE discoveries to ChIP-seq binary scores.")) |>
  kable_styling(position = "center", latex_options = "hold_position")

merged_results |> 
  group_by(binary_score) |>
  summarise(round(mean(seurat_discovery), 2)) |>
  kable(booktabs = TRUE, 
      linesep = "",
      col.names = c("Binary ChIP-seq score", "Proportion Seurat discoveries"),
      caption = paste0("Comparing Seurat discoveries to ChIP-seq binary scores.")) |>
  kable_styling(position = "center", latex_options = "hold_position")
```

```{r, fig.width = 4, fig.height = 3}
merged_results |>
  ggplot(aes(x = sceptre_discovery, y = linear_score)) +
  geom_violin(fill = "dodgerblue") +
  labs(x = "SCEPTRE discovery",
       y = "ChIP-seq linear score")

merged_results |>
  ggplot(aes(x = seurat_discovery, y = linear_score)) +
  geom_violin(fill = "dodgerblue") +
  labs(x = "Seurat discovery",
       y = "ChIP-seq linear score")
```

These results suggest that there is no enrichment of ChIP-seq signal in either the SCEPTRE or Seurat discoveries, regardless of how ChIP-seq signal is measured.

# Conclusions

Below are the answers to the questions posed at the beginning:

- How do people typically determine transcription factor targets? **There isn't just one way of doing this. One lesson learned from the literature is that the window sizes used are larger than our original one (500bp) and are centered on the TSS rather than just upstream of it. Perhaps ChromHMM would help, but no one appears to have used this approach so perhaps neither should we.**
- How do ChIP-seq scores for potential target genes align with those in the hTFtarget database? **The alignment is pretty decent, suggesting that we're not making a big mistake processing the ChIP-seq data.**
- How do SCEPTRE and Seurat IRF1 discoveries align with each other? **The alignment is decent but very imperfect, as Kaishu has found before.**
- How do SCEPTRE and Seurat IRF1 discoveries align with ChIP-seq scores? **There appears to be no enrichment of ChIP-seq signal in the SCEPTRE and Seurat discoveries.**

The last point is the puzzling one. My conjecture is that we are looking at ChIP-seq data from the wrong cell type. The next step would be to look at monocytes treated by IFN-gamma.

\newpage

# Appendix

It seems based on the code below that `join_overlap_left` does take chromosome information into account.
```{r, warning = FALSE, echo = TRUE}
test1 <- GRanges(
  seqnames = "chr1",
  ranges = IRanges(start = 10, end = 20))

test2 <- GRanges(
  seqnames = "chr2",
  ranges = IRanges(start = 10, end = 20),
  metadata = "a")

test3 <- GRanges(
  seqnames = "chr1",
  ranges = IRanges(start = 10, end = 20),
  metadata = "a")


join_overlap_left(test1, test2) # different chromosomes

join_overlap_left(test1, test3) # same chromosome
```