---
title: "IRF1 SCEPTRE vs Seurat in monocytes, with same QC"
output: pdf_document
date: "2023-04-04"
header-includes:
- \usepackage{diagbox}
- \usepackage{multicol}
- \usepackage{float}
---

```{r, echo = FALSE}
knitr::opts_chunk$set(fig.align = 'center', echo = FALSE)
```

```{r, message = FALSE, warning = FALSE, cache = FALSE}
# load libraries
library(tidyverse)
library(plyranges)
library(genomation)
library(biomaRt)
library(GenomicRanges)
library(kableExtra)
library(Hmisc)

# resolve namespace conflicts
library(conflicted)
conflicts_prefer(dplyr::filter)
conflicts_prefer(dplyr::select)
conflicts_prefer(dplyr::rename)
conflicts_prefer(base::intersect)
```

# Introduction

This is a followup analysis of Gene's `IRF1-analysis-v2` writeup, with two main differences: **The ChIP-seq data is from CD14+ monocytes rather than K562 and the Seurat QC is matched to that of SCEPTRE.** Here, we consider two ways of determining transcription factor target genes: those that are in the hTFtarget database and those that have at least one ChIP-seq peak within 5kb of their TSS. These two align reasonably well; see the table below.

```{r, message = FALSE}
data_dir <- .get_config_path("LOCAL_SCEPTRE2_DATA_DIR")

# read in sceptre and seurat results
sceptre_path <- paste0(
  data_dir, "results/papalexi_analysis/",
  "sceptre_full_mrna_results_with_effect_size.rds"
)
seurat_path <- paste0(
  data_dir, "results/papalexi_analysis/",
  "seurat_all_perturbations_results.rds"
)
sceptre_results <- readRDS(sceptre_path)
seurat_results <- readRDS(seurat_path)

# read in K562 ChIP-seq data
stat1_filename <- "GSM1057011_STAT1peak_B.txt"
chipseq_fp <- paste0(data_dir, "data/chipseq/", stat1_filename)
chipseq_data <- read.table(chipseq_fp) 
colnames(chipseq_data) = c("chrom",'start_pos','end_pos',"pval","score",
                           "pos_max_peak","max_peak_height",
                           "rel_pos_max_peak_height","peak_size","mid_point",
                           "peak_to_mid_dist")

# read in genes from database
database_genes <- read_table(paste0(
  data_dir,
  "data/htftarget/dataset_3390.STAT1.target.txt"
)) |>
  suppressWarnings() |>
  pull(target_name)
```

```{r}
chipseq_data = GRanges(
  seqnames = chipseq_data$chrom,
  ranges = IRanges(start = chipseq_data$start_pos, end = chipseq_data$end_pos),
  score = chipseq_data$score,
  peak_position = chipseq_data$pos_max_peak)
```

```{r}
# get TSS for each gene 
gene_names <- sceptre_results |> pull(response_id) |> unique() |> as.character()
ensembl <- useEnsembl(host = 'https://grch37.ensembl.org',
                      biomart = 'ENSEMBL_MART_ENSEMBL', 
                      dataset = "hsapiens_gene_ensembl")
TSS_info <-getBM(attributes=c("external_gene_name", "chromosome_name", "start_position", 
                       "end_position", "strand"),
                 filters=c('external_gene_name'),
                 value = gene_names, mart=ensembl) |>
  filter(chromosome_name %in% c(1:22, "X", "Y")) |>
  mutate(TSS = ifelse(strand == 1, start_position, end_position),
         chromosome_name = paste0("chr", chromosome_name)) |>
  rename(gene = external_gene_name, chr = chromosome_name) |>
  select(gene, chr, TSS)
```


```{r}
# get binary score for each gene
window_width_binary <- 5e3
chipseq_scores_binary <- GRanges(
  seqnames = TSS_info$chr,
  ranges = IRanges(start = TSS_info$TSS-window_width_binary, 
                   end = TSS_info$TSS+window_width_binary),
  gene = TSS_info$gene,
  TSS = TSS_info$TSS) |> 
  join_overlap_left(chipseq_data) |>
  group_by(gene) |>
  summarise(binary_score = any(!is.na(score))) |>
  as_tibble()

# get linear score for each gene
window_width_linear <- 5e4
chipseq_scores_linear <- GRanges(
  seqnames = TSS_info$chr,
  ranges = IRanges(start = TSS_info$TSS-window_width_linear, 
                   end = TSS_info$TSS+window_width_linear),
  gene = TSS_info$gene,
  TSS = TSS_info$TSS) |> 
  join_overlap_left(chipseq_data) |>
  group_by(gene) |>
  summarise(linear_score = sum((window_width_linear - abs(TSS - peak_position))/
                                 window_width_linear)) |>
  as_tibble() |>
  mutate(linear_score = ifelse(is.na(linear_score), 0, linear_score))
```

```{r}
# join the two sets of chipseq scores, and add database information
chipseq_scores <- full_join(chipseq_scores_binary, 
                            chipseq_scores_linear, 
                            by = "gene") |>
  mutate(in_database = gene %in% database_genes)
```

```{r, results = "asis"}
chipseq_scores |> 
  select(in_database, binary_score) |>
  table() |>
  Hmisc::latex(booktabs = TRUE,
               rowlabel="\\diagbox{Database}{ChIP-seq}", 
               file="",
               caption = "Comparing target genes identified based on database and based directly on ChIP-seq data (at least one peak within 5kb).",
               where = "!htbp")
```

```{r, fig.width = 4, fig.height = 3, eval = FALSE}
chipseq_scores |> 
  group_by(in_database) |>
  summarise(round(mean(binary_score), 3)) |>
  kable(booktabs = TRUE, 
        linesep = "",
        col.names = c("In database", "Mean binary score"),
        caption = paste0("Comparing ChIP-seq binary score to database.")) |>
  kable_styling(position = "center", latex_options = "HOLD_position")

chipseq_scores |> 
  group_by(in_database) |>
  summarise(round(median(linear_score), 3)) |>
  kable(booktabs = TRUE, 
        linesep = "",
        col.names = c("In database", "Median linear score"),
        caption = paste0("Comparing ChIP-seq linear score to database.")) |>
  kable_styling(position = "center", latex_options = "HOLD_position")

chipseq_scores |> 
  ggplot(aes(x = in_database, y = linear_score)) +
  geom_violin(fill = "dodgerblue") +
  labs(x = "In database", y = "Linear score")
```

# Compare SCEPTRE and Seurat results with ChIP-seq scores

Let's first look at how the SCEPTRE and Seurat discoveries align with each other.
```{r}
genes_qc <- intersect(sceptre_results |> 
                        filter(grna_group == "IRF1") |>
                        na.omit() |>
                        pull(response_id),
                      chipseq_scores |> pull(gene))

q <- 0.1

# apply BH correction to SCEPTRE IRF1 results
sceptre_results_IRF1 <- sceptre_results |>
  filter(grna_group == "STAT1", response_id %in% genes_qc) |>
  mutate(sceptre_discovery = p.adjust(p_value, method = "BH") <= q) |>
  select(response_id, sceptre_discovery) |>
  rename(gene = response_id)

# apply BH correction to Seurat IRF1 results
seurat_results_IRF1 <- seurat_results |>
  filter(grna_group == "STAT1", response_id %in% genes_qc) |>
  mutate(seurat_discovery = p.adjust(p_val, method = "BH") <= q) |>
  select(response_id, seurat_discovery) |>
  rename(gene = response_id)

# join with chipseq scores
merged_results <- chipseq_scores |>
  filter(gene %in% genes_qc) |>
  left_join(sceptre_results_IRF1, by = "gene") |>
  left_join(seurat_results_IRF1, by = "gene")
```

```{r, results='asis'}
merged_results |> 
  select(sceptre_discovery, seurat_discovery) |>
  table() |>
  Hmisc::latex(booktabs = TRUE,
               rowlabel="\\diagbox{SCEPTRE}{Seurat}", 
               file="",
               caption = "Comparing SCEPTRE versus Seurat discoveries.",
               where = "!htbp")
```

This table suggests that SCEPTRE and Seurat results have decent, but imperfect, agreement. Also note that the total numbers of discoveries made by the two methods are nearly the same. Next, let's look at how the SCEPTRE and Seurat discoveries align with the ChIP-seq target genes (as identified by either the hTFtarget database or directly from the ChIP-seq data).

\newpage

```{r, results='asis'}
add_proportions <- function(tab){
  tab |>
    addmargins(quiet = TRUE, FUN = list(Prop = function(x){
      if(any(x < 1)) NA_character_ else round(x["TRUE"]/sum(x), 3)
    }))
}

merged_results |> 
  select(in_database, sceptre_discovery) |>
  table() |>
  add_proportions() |>
  Hmisc::latex(booktabs = TRUE,
             rowlabel="\\diagbox{Database}{SCEPTRE}", 
             file="",
             caption = "Comparing SCEPTRE to database.", 
             where = "!htbp")

merged_results |> 
  select(in_database, seurat_discovery) |>
  table() |>
  add_proportions() |>
  Hmisc::latex(booktabs = TRUE,
           rowlabel="\\diagbox{Database}{Seurat}", 
           file="",
           caption = "Comparing Seurat to database.",
           where = "!htbp")

merged_results |> 
  select(binary_score, sceptre_discovery) |>
  table() |>
  add_proportions() |>
  Hmisc::latex(booktabs = TRUE,
           rowlabel="\\diagbox{ChIP-seq}{SCEPTRE}", 
           file="",
           caption = "Comparing SCEPTRE to ChIP-seq binary scores.",
           where = "!htbp")

merged_results |> 
  select(binary_score, seurat_discovery) |>
  table() |>
  add_proportions() |>
  Hmisc::latex(booktabs = TRUE,
           rowlabel="\\diagbox{ChIP-seq}{Seurat}", 
           file="",
           caption = "Comparing Seurat to ChIP-seq binary scores.",
           where = "!htbp")
```

The proportions are the proportion of `TRUE` values in each row or column. For example, 0.576 of the genes found by SCEPTRE are marked as IRF1 targets in the database (i.e. SCEPTRE has specificity 0.576). From these tables, we see that **SCEPTRE has slightly lower sensitivity and specificity than Seurat.** We can summarize each 2-by-2 table via its odds ratio (the p-values are all extremely small). The resulting odds ratios are shown below.

```{r, results = 'asis'}
odds_ratios <- matrix(NA, nrow = 2, ncol = 2, 
                      dimnames = list(ground_truth = c("database", "ChIP-seq"), 
                                      method = c("SCEPTRE", "Seurat")))

odds_ratios["database", "SCEPTRE"] <- merged_results |> 
  select(in_database, sceptre_discovery) |>
  table() |>
  fisher.test() %>%
  `$`("estimate")

odds_ratios["database", "Seurat"] <- merged_results |> 
  select(in_database, seurat_discovery) |>
  table() |>
  fisher.test() %>%
  `$`("estimate")

odds_ratios["ChIP-seq", "SCEPTRE"] <- merged_results |> 
  select(binary_score, sceptre_discovery) |>
  table() |>
  fisher.test() %>%
  `$`("estimate")

odds_ratios["ChIP-seq", "Seurat"] <- merged_results |> 
  select(binary_score, seurat_discovery) |>
  table() |>
  fisher.test() %>%
  `$`("estimate")

odds_ratios |>
  round(3) |>
  Hmisc::latex(booktabs = TRUE,
           rowlabel="\\diagbox{Ground truth}{Method}", 
           file="",
           caption = "Enrichment odds ratios, comparing to database and our ChIP-seq target assignments.",
           where = "!htbp")
```

```{r, eval = FALSE}
merged_results |> 
  group_by(sceptre_discovery) |>
  summarise(round(median(linear_score), 3)) |>
  kable(booktabs = TRUE, 
      linesep = "",
      col.names = c("SCEPTRE discovery", "Median linear ChIP-seq score"),
      caption = paste0("Comparing SCEPTRE discoveries to ChIP-seq linear scores.")) |>
  kable_styling(position = "center", latex_options = "HOLD_position")

merged_results |> 
  group_by(seurat_discovery) |>
  summarise(round(median(linear_score), 3)) |>
  kable(booktabs = TRUE, 
      linesep = "",
      col.names = c("Seurat discovery", "Median linear ChIP-seq score"),
      caption = paste0("Comparing Seurat discoveries to ChIP-seq linear scores.")) |>
  kable_styling(position = "center", latex_options = "HOLD_position")
```

These results suggest that **there is nontrivial (roughly two-fold) enrichment of ChIP-seq signal in both the SCEPTRE and Seurat discoveries, presumably because CD14+ cells are a better match to THP1 cells. Unfortunately, the Seurat discoveries have slightly higher enrichment.** This result is somewhat contradictory to our results on control data, where SCEPTRE found fewer false positives and more true positives than Seurat.

\newpage

# Appendix

Another way of measuring the amount of ChIP-seq signal near a gene is the linear score proposed by [Sikora-Wohlfeld et al, 2013](https://journals.plos.org/ploscompbiol/article/file?id=10.1371/journal.pcbi.1003342&type=printable). In this approach, the relative distances of ChIP-seq peaks to the TSS are summed, restricting attention to a 50kb window centered on the TSS. Below is the distribution of the linear IRF1 ChIP-seq scores across genes:

```{r, fig.width = 4, fig.height = 3, message = FALSE}
chipseq_scores |> 
  ggplot(aes(x = linear_score)) +
  geom_histogram(color = "black") +
  labs(x = "ChIP-seq linear score") +
  theme(axis.title.y = element_blank())
```

We see that there are modes at 0 (no peaks within the window width) and 1 (one peaks near the TSS). There is also a long right tail.

Now, let's see the distributions of these linear scores for genes detected by SCEPTRE and Seurat.

```{r, fig.width = 6, fig.height = 3}
cowplot::plot_grid(
  merged_results |>
    ggplot(aes(x = sceptre_discovery, y = linear_score)) +
    geom_violin(fill = "dodgerblue") +
    geom_boxplot(width = 0.1, outlier.colour = NA, fill = NA) +
    labs(x = "SCEPTRE discovery",
         y = "ChIP-seq linear score"),
  merged_results |>
    ggplot(aes(x = seurat_discovery, y = linear_score)) +
    geom_violin(fill = "dodgerblue") +
    geom_boxplot(width = 0.1, outlier.colour = NA, fill = NA) +
    labs(x = "Seurat discovery",
         y = "ChIP-seq linear score"),
  nrow = 1
)
```

Again, we see some nontrivial enrichment for both SCEPTRE and Seurat, without a significant difference apparent between the two methods.




